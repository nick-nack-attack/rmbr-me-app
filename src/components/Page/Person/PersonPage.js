// component shows a person and their rmbrs
import React, { useState, useContext, useEffect } from 'react';
import { useHistory } from 'react-router-dom';

// components
import RmbrList from '../../List/Rmbr/RmbrList';
import EditPersonForm from "../../Form/Person/Edit/EditPersonForm";
import { Button } from '../../Utils/Utils';

// utils
import PrettyDate, { findRmbrByPersonId } from "../../Utils/helpers";

// contexts / services
import RmbrmeContext from "../../../contexts/RmbrmeContext";
import AppApiService from "../../../services/app-api-service";
import { AppContext } from "../../../contexts/AppContext";

// styling
import './PersonPage.scss'; 
import PersonMenu from '../../Menu/Person/PersonMenu';

const PersonPage = props => {

    const [ showForm, setShowForm ] = useState(false);
    const [ id, setId ] = useState(props.match.params.person_id);
    const [ error, setError ] = useState(null);
    const [ rmbrs, setRmbrs ] = useState([]);
    const [ person, setPerson ] = useState({});
    const history = useHistory();

    let context = useContext(RmbrmeContext);
    let appContext = useContext(AppContext);

    useEffect(() => {
        AppApiService.getPersonByPersonId(id)
            .then(person => {
                context.setSelectedPerson(person);
                setPerson(person);
            })
            .catch(err => {
                context.setError(err);
            });
    }, [id]);

    useEffect(() => {
        let contextRmbrs = appContext.state.rmbrs;
        setRmbrs( contextRmbrs.filter(r => r.person_id == props.match.params.person_id) );
    }, [ appContext ])

    const handleHideEditForm = () => {
        setShowForm(false);
    };

    const handleShowEditForm = () => {
        setShowForm(true);
    };

    const handleDelete = () => {

    };

    const renderHeader = () => {
        return (
            <div className="person-header"> 
            <h2>
                { person.person_name }
            </h2>
            <h3>
                { person.type_of_person } since { PrettyDate(person.date_created || new Date()) }
            </h3>
            <PersonMenu
                handleEditClick={ () => handleShowEditForm() }
                handleDeleteClick={ e => handleDelete(e) }
            />
        </div>
        );
    };

    const renderEdit = () => {
        return (
            <EditPersonForm 
                person={ person }
                handleCancelClick={ () => handleHideEditForm() }
            />
        );
    };

    const handleConfirmDelete = (e, id) => {
        if ( window.confirm(`Are you sure you want to delete this person?`)) {
            e.preventDefault();
            setError(null);
            AppApiService.deletePerson(id)
                .then(() => {
                    appContext.dispatch({
                        type: 'refetch'
                    });
                })
                .then(() => {
                    history.push('/');
                })
                .catch(err => {
                    setError(err);
                })
        } 
    };

    return (

        <div>

            { showForm ? renderEdit() : renderHeader() }

            <Button
                label='Back'
                className='back_to_root_button'
                onClick={ () => history.push('/') }
            />
        
            <div>
                <RmbrList 
                    id={id}
                    person={ person }
                    rmbrs={ rmbrs }
                />
            </div>

        </div>

        /*
        <div key={person_id}>
                <header>
                    { !this.state.editFormDisplayed
                        ?   <>
                                <div className='person_page_header'>
                                    <h2>
                                        { person_name }
                                        <span className='edit_title_inline_button'>
                                            <button
                                                onClick={() => this.handleEditPerson()}
                                            >
                                                <FontAwesomeIcon icon='pen' />
                                            </button>
                                                <div className='person_page_delete_button_div'>
                                                    <button
                                                        className='list_delete_button'
                                                        onClick={(e) => this.handleDeletePerson(e)}
                                                    >
                                                        <FontAwesomeIcon icon='trash-alt' />
                                                    </button>
                                                </div>
                                        </span>
                                    </h2>
                                    <h3> {type_of_person}</h3>
                                    <p>Added {PrettyDate(date_created)}</p>
                                </div>
                            </>
                        :   <EditPersonForm
                                person_id={person_id}
                                person_name={person_name}
                                type_of_person={type_of_person}
                                onHideEditForm={e => this.onHideEditForm(e)}
                            />
                    }
                </header>
                <div>
                    { error
                        ?   <p className='red'>Could not load Rmbrs list</p>
                        :   <RmbrList
                            person_id={person_id}
                            rmbrList={this.context.rmbrList}
                            />
                    }
                <div>
                <div>
                    { this.renderBackButton() }
                </div>
            </div>
            </div>
            </div>
            */
    );

};

export default PersonPage;