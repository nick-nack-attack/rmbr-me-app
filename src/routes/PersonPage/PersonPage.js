import React, { Component } from 'react';
import RmbrApiService from "../../services/rmbr-api-service";
import RmbrList from '../../components/RmbrList/RmbrList'
import EditPersonForm from "../../components/EditPersonForm/EditPersonForm";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { Button } from "../../components/Utils/Utils";
import RmbrmeContext from "../../contexts/RmbrmeContext";
import PrettyDate from "../../components/Utils/Utils";


export default class PersonPage extends Component {

    constructor(props) {
        super(props);
        this.state = {
            editFormDisplayed: false,
            person_id: this.props.match.params.person_id,
            onDeletePerson: () => {}
        }
    };

    static contextType = RmbrmeContext;

    componentDidMount() {
       RmbrApiService.getPersonByPersonId(this.state.person_id)
           .then( person => {
               this.context.setSelectedPerson(person)
           })
           .catch(err => {
               this.context.setError(err)
            })
    };

    onHideEditForm = e => {
        this.setState({
            editFormDisplayed: false
        })
    };

    handleDeletePerson = e => {
        e.preventDefault()
        const person_id = this.props.match.params.person_id;
        this.setState({error: null});
        RmbrApiService.deletePerson(person_id)
            .then(() => {
                this.context.deletePerson(person_id)
            })
            .then(() => this.props.history.push('/'))
            .catch(res => {
                this.setState({
                    error: res.error
                })
            })
    };

    handleEditPerson = () => {
        this.setState({
            editFormDisplayed: true
        })
    };

    renderBackButton = () => {
        return (
            <Button
                className='back_to_root_button'
                onClick={() => this.props.history.push('/')}
            >
                <FontAwesomeIcon id='arrow-left' icon='arrow-left'/>
                Back
            </Button>
        )
    }

    render() {
        const person_id = this.props.match.params.person_id;
        const person_name = this.context.selected_person.person_name
        const type_of_person = this.context.selected_person.type_of_person || '';
        const date_created = this.context.selected_person.date_created || new Date();
        const { error } = this.context;

        return (
            <div key={person_id}>
                <header>
                    { !this.state.editFormDisplayed
                        ?   <>
                                <div className='person_page_header'>
                                    <h2>
                                        { person_name }
                                        <span className='edit_title_inline_button'>
                                            <button
                                                onClick={() => this.handleEditPerson()}
                                            >
                                                <FontAwesomeIcon icon='pen' />
                                            </button>
                                                <div className='person_page_delete_button_div'>
                                                    <button
                                                        className='list_delete_button'
                                                        onClick={(e) => this.handleDeletePerson(e)}
                                                    >
                                                        <FontAwesomeIcon icon='trash-alt' />
                                                    </button>
                                                </div>
                                        </span>
                                    </h2>
                                    <h3> {type_of_person}</h3>
                                    <p>Added {PrettyDate(date_created)}</p>
                                </div>
                            </>
                        :   <EditPersonForm
                                person_id={person_id}
                                person_name={person_name}
                                type_of_person={type_of_person}
                                onHideEditForm={e => this.onHideEditForm(e)}
                            />
                    }
                </header>
                <div>
                    { error
                        ?   <p className='red'>Could not load Rmbrs list</p>
                        :   <RmbrList
                            person_id={person_id}
                            rmbrList={this.context.rmbrList}
                            />
                    }
                <div>
                <div>
                    { this.renderBackButton() }
                </div>
            </div>
            </div>
            </div>
        )
    }
}