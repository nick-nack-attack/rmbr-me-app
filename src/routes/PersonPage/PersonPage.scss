@import '../../themes/v2';

div.person-header {
    text-align: left;
    margin-left: auto;
    margin-right: auto;
    max-width: 470px;
    padding: 30px 40px;
    color: black;
    &>h2 {
        font-size: 30px;
        font-weight: bold;
    }
    &>h3 {
        font-size: 20px;
    }
}

.options-div {
    float: right;
    margin-top: -20;
}
