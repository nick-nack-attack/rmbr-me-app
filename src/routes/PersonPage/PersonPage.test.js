import React from 'react';
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import PersonPage from "./PersonPage";

describe(`PersonPage component`, () => {

    it('renders PersonPage by default', () => {
        const wrapper = shallow(<PersonPage/>)
        expect(toJson(wrapper)).toMatchSnapshot()
    })

})