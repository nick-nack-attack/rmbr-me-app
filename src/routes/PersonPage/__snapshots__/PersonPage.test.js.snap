// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonPage component renders PersonPage by default 1`] = `
<testContext>
  <PersonPage
    context={
      Object {
        "$$typeof": Symbol(react.context),
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "_calculateChangedBits": null,
          "_context": [Circular],
        },
        "Provider": Object {
          "$$typeof": Symbol(react.provider),
          "_context": [Circular],
        },
        "_calculateChangedBits": null,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": Object {
          "selected_person": Object {},
        },
        "_currentValue2": Object {
          "selected_person": Object {},
        },
        "_threadCount": 0,
      }
    }
    match={
      Object {
        "isExact": true,
        "params": Object {
          "person_id": 1,
        },
        "path": "",
        "url": "",
      }
    }
  />
</testContext>
`;
